java/ql/src/Telemetry/AutomodelFrameworkModeCharacteristics.qll

```diff
@@ -0,0 +1,319 @@
+/**
+ * For internal use only.
+ */
+
+private import java
+private import semmle.code.Location as Location
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.TaintTracking
+private import semmle.code.java.security.PathCreation
+private import semmle.code.java.dataflow.ExternalFlow as ExternalFlow
+private import semmle.code.java.dataflow.internal.FlowSummaryImpl as FlowSummaryImpl
+private import semmle.code.java.security.ExternalAPIs as ExternalAPIs
+private import semmle.code.java.Expr as Expr
+private import semmle.code.java.security.QueryInjection
+private import semmle.code.java.security.RequestForgery
+private import semmle.code.java.dataflow.internal.ModelExclusions as ModelExclusions
+import AutomodelSharedCharacteristics as SharedCharacteristics
+import AutomodelEndpointTypes as AutomodelEndpointTypes
+
+Callable getCallable(DataFlow::ParameterNode e) { result = e.getEnclosingCallable() }
```

Comment:
QLDoc would be great. I'd also like to see this shared/parameterised between application and framework mode, but we can do that in a follow-up PR. Perhaps that would be a `getCallable(Endpoint e)` predicate as part of the `CandidateSig` module signature, or a member predicate we expect as part of the `Endpoint` class signature.

Reply:
Formalising this a bit better in application mode is what I was hoping to do (assuming I'll work on that). For now, I've added qldoc (good point 👍), and moved it to the candidate signature module as an `additional` predicate to emphasise that it's really part of the mode-specific implementation. The qldoc contains a note that every mode should implement this.